validator.go

```diff
@@ -117,16 +117,31 @@ func ValidateAll(xmlReader io.Reader) []error {
 
 // bufio implements a ByteReader but we explicitly don't want any buffering
 type byteReader struct {
-	r io.Reader
+	r   io.Reader
+	err error
 }
 
 func (r *byteReader) ReadByte() (byte, error) {
-	p := make([]byte, 1)
-	_, err := r.r.Read(p)
-	return p[0], err
+	var p [1]byte
+	n, err := r.Read(p[:])
+
+	if n > 0 {
+		// store the error if any :
+		r.err = err
+		// return the byte read from the reader
+		return p[0], nil
+	}
+
+	return 0, err
 }
 
 func (r *byteReader) Read(p []byte) (int, error) {
+	if r.err != nil {
+		err := r.err
+		r.err = nil
+		return 0, err
+	}
```

Comment:
Why is this part and storing the error in `ReadByte` needed? Wouldn't `r.r.Read(p)` return 0 and an error anyway if we already encountered an EOF before?

Reply:
In ReadByte : we hide *any* error, not only EOF.

1. I chose to return that error once, to not completely hide it from the calling code.

Alternatives could be :

2. discard that error (as you suggest) : if 1 byte is read in ReadByte, return that byte and completely discard that error, let the next call to Read() return an error if need be ;
3. make some extra checks on the error : for example, if an error is temporary (as in `net.Error` for example), do not store it

I'm mentioning 3. just to give an example of something more elaborate, but I wouldn't go for it,  
I'm fine with 1. and 2.

Do you have an opinion ?

Reply:
Considering this is an absolutely critical piece of code, I'd highly prioritize keeping it as simple as possible -- for readability, auditability, and to reduce the potential for introducing new bugs. Option 2. would remove several lines of code from this change, so that's the one I'd go with. Counter-arguments are absolutely welcome and appreciated, though.

Reply:
I'm fine with dropping the stored error, I'll mention one last possibility : do not wrap the reader in a `byteReader{}` and let the standard `xml.NewDecoder()` impleme,ntation add its `bufio.Reader()` wrapper.

I see the benefit on resource usage : most `io.Reader`s used in these context will already have some form of buffering so adding another buffer on top is probably not very useful ;  
did you have some extra arguments in mind when adding this custom wrapper, though ?

Reply:
Yes, the custom wrapper gets rid of the buffering that would otherwise be added by the decoder. We explicitly don't want any, because the same reader is used in multiple decoder instances and want to be sure we continue reading where we previously stopped decoding, not where the buffer ended.